### PR TITLE
Class existence resource

### PR DESCRIPTION
**Pull request by fabpot**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | see #20094 |
| License | MIT |
| Doc PR | n/a |

### CODE REVIEW

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Config\Resource;
+
+/**
+ * ClassExistenceResource represents a class availability.
```
Comment (wouterj) [inline src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: Let's replace `availability` with `existence`, to be consistent with the class name and the `class_exists` function name.

File: src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        eval(<<<EOF
+namespace Symfony\Component\Config\Tests\Fixtures;
+
+class BarClass
+{
+}
+EOF
+        );
+
+        $this->assertFalse($res->isFresh(time()));
```
Comment (wouterj) [inline src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php]: Afaik, this should be `assertTrue`. In table form (horizontal: current state, vertical: stored state)

| | Class exists | Class does not exists |
| --- | --- | ---|
| **Class exists** | Not fresh | Fresh |
| **Class does not exists** | Fresh | Not fresh |

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php [hunk 3]
```diff
@@ -72,15 +76,18 @@ public function process(ContainerBuilder $container)
             $container->getDefinition('twig.extension.assets')->addTag('twig.extension');
         }
 
-        if (class_exists('Symfony\Component\Yaml\Parser')) {
+        $container->addResource(new ClassExistenceResource(YamlParser::class));
+        if (class_exists(YamlParser::class)) {
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php]: Related to [this comment](https://github.com/symfony/symfony/commit/79efb4c8414ea459d1ad678ed79ba469e3597d51#commitcomment-19258739), just asking if could we replace this:

```php
if (class_exists(YamlParser::class)) {
    $container->getDefinition('twig.extension.yaml')->addTag('twig.extension');
}
```

by this?

```php
if (!class_exists(YamlParser::class)) {
    $container->removeDefinition('twig.extension.yaml');
}
```

Comment (nicolas-grekas) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php]: The difference is that if one uses the `twig.extension.yaml` directly, the proposed version will break.
I think the current approach is a more open thus a better practice, even if in this very case using the extension directly is so unlikely.

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: done

File: src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php (hunk 2 above)
Comment (fabpot) [inline src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php]: The freshness is all about the state change. The resource is fresh if there is no change in class existence, so no need to refresh the cache. Not fresh means that the cache must be recalculated, so when the state changes. So, the table should be:

| | Class exists | Class does not exists |
| --- | --- | ---|
| **Class exists** | Fresh | Not fresh |
| **Class does not exists** | Not fresh | fresh |

Comment (wouterj) [inline src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php]: Oh, sorry. Completely messed up the definition of "fresh"...

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    private $resource;
+    private $exists;
+
+    /**
+     * @param string $resource The fully-qualified class name
+     */
+    public function __construct($resource)
+    {
+        $this->resource = $resource;
+        $this->exists = class_exists($resource);
```
Comment (stof) [inline src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: we should use ``class_exists() || interface_exists() || trait_exists()`` to make it usable for interfaces and traits too

**Pull request merged**